server/data/data-library/src/main/java/org/apache/james/rrt/lib/CanSendFromImpl.java

```diff
@@ -79,29 +83,38 @@ private boolean emailIsAnAliasOfTheConnectedUser(Username connectedUser, Usernam
             .anyMatch(alias -> alias.equals(connectedUser));
     }
 
-    private Stream<Username> relatedAliases(Username user) throws RecipientRewriteTableException {
-        return Stream.concat(
-            Stream.of(user),
+    private Stream<Username> relatedAliases(Username user) {
+        return recursivelyResolve(user, Throwing.<Username, Stream<Username>>function(targetUser ->
             recipientRewriteTable
-                .listSources(Mapping.alias(user.asString()))
+                .listSources(Mapping.alias(targetUser.asString()))
                 .map(MappingSource::asUsername)
-                .flatMap(OptionalUtils::toStream)
+                .flatMap(OptionalUtils::toStream)).sneakyThrow()
         );
     }
 
     private Stream<Domain> relatedDomains(Username user) {
         return user.getDomainPart()
-            .map(Throwing.function(this::fetchDomains).sneakyThrow())
+            .map(this::fetchDomains)
             .orElseGet(Stream::empty);
     }
 
-    private Stream<Domain> fetchDomains(Domain domain) throws RecipientRewriteTableException {
-        return Stream.concat(
-          Stream.of(domain),
+    private Stream<Domain> fetchDomains(Domain domain) {
+        return recursivelyResolve(domain, Throwing.<Domain, Stream<Domain>>function(targetDomain ->
           recipientRewriteTable
-              .listSources(Mapping.domain(domain))
+              .listSources(Mapping.domain(targetDomain))
               .map(MappingSource::asDomain)
-              .flatMap(OptionalUtils::toStream)
+              .flatMap(OptionalUtils::toStream)).sneakyThrow()
         );
     }
+
+    private <T> Stream<T> recursivelyResolve(T seed, Function<T, Stream<T>> generator) {
```

Comment:
I will read 10 times this function before commenting it :)

Reply:
I think you might need `distinct()`

Reply:
In which case?

Reply:
with RESOLUTION_RECURSION_LEVEL = 2 you may not have a many loops (still possible) but if we decide to use more than 2, you can have the same entry many times and it may be amplified with each `unfold` iteration.
The seems correct right now but we could avoid generating duplicates to make it more efficient.